Set up packaging to allow standard installation

diazona opened the pull request:
This PR adds the metadata needed to build distributable packages (sdist and wheel) for pypipe and a Github Actions workflow that - after suitable configuration - will upload the package to PyPI each time a tag is made.

I'll come back to this and add instructions for how to do that configuration, but at the moment I don't have time for that, so I'm keeping this as a draft for now. I figured I should just post my work in progress so other people have some idea of what I'm planning, and so if this would clash with other PRs people would like to propose, we can cleanly work around any conflicts.

Closes #4

gwerbin commented:
@diazona what do you think about adding a couple of Hatch environments?

```toml
[tool.hatch.envs.dev]
dependencies = [
  "twine"
]

[tool.hatch.envs.dev.scripts]
publish-test = "twine upload -r testpypi dist/*"
```

I know there isn't yet a test suite, but this also sets a good precedent for adding test dependencies later using Hatch (which also facilitates [version matrix testing](https://hatch.pypa.io/latest/environment/#matrix)).

@bugen if you're not already familiar with the packaging & publishing workflow, you'll probably want to work through the [packaging and distribution tutorial](https://packaging.python.org/en/latest/tutorials/packaging-projects/) if you're not already familiar with it.

diazona commented:
I don't have any objections to adding a dev environment in principle. I had just removed everything nonessential from the generated hatch config file (so it would be quick and simple to get merged) but I wouldn't mind adding it back.

However... I'm a _little_ reluctant to use that environment to publish packages. (I know your comment just shows the test PyPI version, but I would think it makes sense to do both test and real uploads the same way.) I had wanted to set up [trusted publishing](https://docs.pypi.org/trusted-publishers/) since as far as I know that's currently the most secure (for some definition of "secure") and often-recommended way to publish packages to PyPI or Test PyPI, and I thought it might be more confusion than it's worth to introduce multiple different ways to publish. I do recognize that trusted publishing is more complicated to set up than using Twine, but it's the sort of thing you can do once and then basically forget about it.  Nevertheless, if you feel strongly about it (especially @bugen as the maintainer), I'm fine going along with that suggestion.

For context @bugen, the main difference aside from what was already explained is this:

- If you use trusted publishing to upload packages to PyPI, Github will build and upload the files to PyPI for you. You can choose when to make this happen: it could be automatically triggered whenever you make a tag, for example, or you can do it manually by clicking a button. But it does require several different things to be configured. (I'll provide instructions)
- If you use twine to upload packages to PyPI, you have to run the command yourself. You can do it any time you want, but you have to manually check some things, like that the version of the package you're uploading is the one you intend to. It doesn't require much configuration other than creating an account on PyPI.

gwerbin commented:
Nope, I have no attachment to Twine at all. Trusted publishing is a great idea here. No need for the extra Hatch stuff that way, either.

bugen commented:
@diazona @gwerbin
Thank you for the various suggestions! I will study the process of publishing to PyPI.

It seems there are two options, a more secure and reliable method known as "trusted publishing," and a simpler method using "twine". The setup for "trusted publishing" might be a bit complex, but it's a one-time effort. I believe we should choose the more reliable method - "trusted publishing".

@diazona 
> I'll come back to this and add instructions for how to do that configuration, but at the moment I don't have time for that, so I'm keeping this as a draft for now.

There's no need to rush. Let's proceed at a comfortable pace. Thank you for your support!

@gwerbin
> I know there isn't yet a test suite,
I recognize the need for test automation. My friend scolded me for not having tests... I can't promise immediate readiness, but I also want to make progress on the testing aspect.

diazona commented:
Sounds good! And thanks for your patience. I know things can go very quickly when a new project is getting started, especially one with so much interest as this, and I would hate to block anyone else's work by taking too long with my pull request.

I was actually in the middle of writing a comment when you posted, so let me just finish that up. Basically, these are a couple things you'll need to configure, either before or after you merge my pull request.

# Release triggering

In the past, I suppose that publishing a new version of your project was as simple as editing the `__version__` constant in the source code, but now that you're going to be putting your project on PyPI, that won't be enough. You have to actually tell Github to build a new version and upload it to PyPI. The question is, how do you want to do that? There are a few different ways, but the most likely options are:

- When you want to release a new version, you create a Git tag named after the version, push that tag to Github, and Github Actions will automatically build the package and upload it to PyPI for you. For example, when you want to release version 0.3.0 of your project, the tag would be named `v0.3.0` or `0.3.0`. This will require some changes in the code: we'll have to remove the `__version__` constant from the source code, and add some configuration that tells it to determine the version number from the Git tag. (If you want to do this, I'll include those changes in this pull request.)
- Or, when you want to build and upload a new version, you would manually run a workflow. This involves going to your project's [release action page](https://github.com/bugen/pypipe/actions/workflows/release.yml) (which doesn't exist yet, but it will after you merge this PR) and clicking a button. Then Github Actions will automatically build the package and upload it to PyPI for you. If you take this approach, you need to be sure to update the version in the source code and commit that before you run the workflow, but you can (or I can help you) set up an automated test to check that.

There are various ways to tweak this to your liking, and I'd be happy to explain more if you have specific ideas about how you want it to work, but to keep it simple, I think it's best to start with one of these two or something similar. You can always change it later. For what it's worth, I use the tag method (the first one) in my own projects, because I like to have exactly one commit that is clearly identified as each specific version. The only drawback of that is that it's a little more complicated to set up than the other version, but again, it's the sort of thing you can do once and then pretty much forget about.

Regardless of which way you choose to trigger releases, I'll need to make some changes to this PR to encode your choice, so let me know how you'd like to proceed. (Or, of course, we can discuss it further if you have questions.)

# Trusted publishing configuration

You'll also need to set up trusted publishing. This applies regardless of which way you choose to trigger releases.

I'd suggest reading [PyPI's documentation about it](https://docs.pypi.org/trusted-publishers/) if you haven't already done so. It sounds like you can [use trusted publishing to create your project in PyPI](https://docs.pypi.org/trusted-publishers/creating-a-project-through-oidc/). This should be what you need to do - although I might have missed something, so if you notice any differences between what I'm saying and the official documentation, please call them out:

1. Create an account on [PyPI](https://pypi.org/), if you haven't already.
2. Follow the instructions from that last link. When you follow the instructions, you would need to enter these values into the "Add a new pending publisher" form:

    - PyPI Project Name: `pypipe`
    - Owner: `bugen`
    - Repository name: `pypipe`
    - Workflow name: `release.yml`
    - Environment name: `pypi`

3. (optional but recommended) Create an account on [Test PyPI](https://test.pypi.org)
4. (if you did step 3) Follow the instructions from the link but on Test PyPI instead of regular PyPI. This time, you need to enter these values into the "Add a new pending publisher" form:

    - PyPI Project Name: `pypipe`
    - Owner: `bugen`
    - Repository name: `pypipe`
    - Workflow name: `release.yml`
    - Environment name: `test-pypi`

5. Go to [your Github project environment settings](https://github.com/bugen/pypipe/settings/environments) and [create an environment](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#creating-an-environment) named `pypi`. Note that the name of the environment you create has to match _both_:

    - The value you entered into the "Environment name" field from the publisher form on PyPI
    - The `environment` value in the file `.github/workflows/release.yml` which I am adding in this PR

    A page will come up that shows you a bunch of settings, but you can leave all the settings as they are, at least for now. (There are a few you might want to change, but you can always do that later, and they're not essential.)
6. (if you did steps 3 and 4) Go back to [the environment settings page](https://github.com/bugen/pypipe/settings/environments) and create another environment named `test-pypi`. Again, the name of the environment has to match the values on PyPI and in `release.yml`. And again, you can leave the settings alone for now.

After all these steps are done, and you've merged this pull request, then you should be all set to make a release.

# Review commit messages

This isn't a configuration thing :smile: I just wanted to let you know that I updated this pull request with more descriptive commit messages that explain what I'm doing at each step. I'd suggest you take a look at the commits and see if the descriptions make sense - if not, feel free to ask and I'll be happy to explain anything in greater detail.

diazona commented:
Oh, and since I just noticed you made [a release on Github](https://github.com/bugen/pypipe/releases/tag/v0.2.0), I should mention: it's also possible to make Github build and upload your code to PyPI whenever you make a release on Github. That's another alternative to doing it whenever you push a tag, or doing it manually.

If you're _really_ curious about the advantages and disadvantages of various ways of doing this, you could look at e.g. [this discussion](https://github.com/diazona/setuptools-pyproject-migration/issues/88#issuecomment-1748214486) from one of my projects where we were discussing the best way to trigger releases. It could give you a sense of what can happen when you choose one approach vs another. But that's getting into details that you really don't need to deal with yet, and honestly that you might never have to deal with. (For the record, we have not yet figured out the best way to do it for that project :sweat_smile:)

bugen commented:
@diazona
Thank you for the detailed explanation; it's really helpful.
I'll create PyPI and TestPyPI accounts promptly. Regarding 'Release triggering,' I'll give it some more thought and respond later.

By the way, it seems that there is already a package named 'pypipe' on PyPI. Is it okay to use the same package name even if it's already in use?

https://pypi.org/project/pypipe/
https://pypi.org/project/python-pypipe/

diazona commented:
Sounds good :+1: 

> By the way, it seems that there is already a package named 'pypipe' on PyPI. Is it okay to use the same package name even if it's already in use?

Oh, sorry, I didn't realize the name was taken. For some reason it didn't come up when I searched for it.

Unfortunately you cannot use a package name that's already in use, so you'll have to choose a different name for your PyPI project. (This does also affect how pypipe can be invoked with pipx, because pipx works best if the name of the script you invoke matches the name of the project on PyPI, but we can deal with that later.)

bugen commented:
@diazona 

```sh
$ pip install pypipe
ERROR: Could not find a version that satisfies the requirement pypipe (from versions: none)
ERROR: No matching distribution found for pypipe
```

'pypipe' might still be available. I'll try to register it. If it's not possible, I'll think of another name.

bugen commented:
Hi @diazona,
I have set up PyPI, TestPyPI, and GitHub Environment according to [your instructions](#issuecomment-1782353481).

PyPI Pending publishers
```
PyPI Project Name: pypipe-ppp
Owner: bugen
Repository Name: pypipe
Workflow name: release.yml
Environment name: pypi
TestPyPI Pending publishers
```

TestPyPI Pending publishers
```
TestPyPI Project Name: pypipe-ppp
Owner: bugen
Repository Name: pypipe
Workflow name: release.yml
Environment name: test-pypi
```

Unfortunately, the name "pypipe" was not available on PyPI. Instead, I have used the name "pypipe-ppp."

diazona commented:
Sounds good, thanks! I'll have to adjust the configuration to account for that name; I'll get back to you when I do. (The fact that the name on PyPI doesn't match the name of the script complicates things just a tiny bit, but hopefully it shouldn't take me too long.)

diazona commented:
OK I did some research and testing and (somewhat surprisingly) I _think_ only a couple minor changes are necessary to handle the package's name on PyPI being `pypipe-ppp`. I pushed new commits with those changes. (I'll want to combine these new commits with earlier ones before this PR gets merged, but I will come back to that later.) If I've forgotten anything, hopefully someone will point it out or it will show up as an error when you try to publish a release to PyPI.

Now I think the only thing I'm waiting on before I can make this PR ready to merge is your opinion on how to trigger releases. No rush, though; take as much time as you like to think about that. I just wanted to make sure it's clear what still needs to be done.

bugen commented:
@diazona Thanks!

> Now I think the only thing I'm waiting on before I can make this PR ready to merge is your opinion on how to trigger releases. 

I would like to start with manual execution at first. I am considering the following release process for pypipe:

1. Create a Git tag and push it to Github.
2. Create a Github release from the Git tag.
3. Create a PyPI release from the same tag.

In the future, we may automate steps 2 and 3 completely from tag creation, but for now, we will manually execute steps 2 and 3.
The following content may include things unrelated to this PR, but there might be parts that could affect your work, so I wanted to share it.

Versioning:
* Increment the version with each new release: 0.2.0 -> 0.3.0 -> 0.4.0 ... -> 1.0.0 -> 1.1.0
* The third digit can be used for urgent releases, such as addressing critical security or bug fixes: 0.3.1

Changelog and Documentation:
* Generate Github Changelog automatically from past commits, with manual adjustments as needed.
* PyPI Changelog should be the same as Github one.
* If content needs to be separated between PyPI and Github Changelog, manually modify PyPI Changelog shortly after release.
* Generate PyPI documentation from Github README.

bugen commented:
commit 5ca17c151d219c30f5b0c7ca91cdad1ac6e33f8b

To run it with pipx run, it's necessary to create an executable file with the same name as the package, which is `pypipe-ppp`. However, it's possible to create the same executable file with a different name, like `pypipe`. In that case, it would be a good idea to add `ppp = "pypipe:main"`. 
The 'ppp' package on PyPI is not a CLI tool, but the other 'pypipe' is a CLI tool and serves a similar purpose to our 'pypipe'. So it's essential to include a note in the README to avoid confusing 'pypipe' and 'pypipe-ppp'.

diazona commented:
Got it, thanks for replying.

> I would like to start with manual execution at first. I am considering the following release process for pypipe:
> 
>  1. Create a Git tag and push it to Github.
>  2. Create a Github release from the Git tag.
>  3. Create a PyPI release from the same tag.
> 
> In the future, we may automate steps 2 and 3 completely from tag creation, but for now, we will manually execute steps 2 and 3.

If I understand correctly what you have in mind, this may not be possible in _exactly_ the way you're thinking about it, but we can get pretty close. What I would suggest is option 1 below:

1. When you create a Github release, the Github actions workflow will get triggered and will automatically build the packages (sdist and wheel) for later uploading. However, Github will prompt you to approve the deployment before uploading those packages to PyPI. If you don't want that release to go on PyPI, you can decline to approve it.

    In order to set this up, you'll need to enable the "Required reviewers" [deployment protection rule](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#required-reviewers) in the `pypi` environment on [the Github environment settings page](https://github.com/bugen/pypipe/settings/environments). You'll also have to add yourself to the list of people who can approve workflow runs. Optionally, you can do the same for the `test-pypi` environment if you want (but then you'll get prompted to approve twice for each release, once for Test PyPI and once for real PyPI).
2. Another option is that you manually run the workflow, and when you do so, you manually give it the name of the tag to build from. This way, no packages get built at all unless you choose to run the workflow.

    Personally, I don't think there is much benefit in doing it this way. But if you like, I can try to set it up.

> Versioning:
> 
> * Increment the version with each new release: 0.2.0 -> 0.3.0 -> 0.4.0 ... -> 1.0.0 -> 1.1.0
> * The third digit can be used for urgent releases, such as addressing critical security or bug fixes: 0.3.1

Makes sense :+1: I don't think this will have any impact on what I'm doing in this PR.

I guess it would be possible to set up some automatic process (such as a Github Actions workflow) to automatically increment the version number in the way you want, but that can be done later, and it is probably a bit complicated. I think it's not worth it at this time.

> Changelog and Documentation:
> 
> * Generate Github Changelog automatically from past commits, with manual adjustments as needed.

Makes sense :+1: You can do this when you create the Github release.

> * PyPI Changelog should be the same as Github one.
> * If content needs to be separated between PyPI and Github Changelog, manually modify PyPI Changelog shortly after release.

Well... the thing about that is, PyPI does not actually host a changelog at all. What people can see on PyPI is a list of versions released (the "Release history"), but that list is automatically generated based on the files you upload to PyPI. You don't get to attach change notes to those releases.

Or, you can host a changelog somewhere else and include a link to it on PyPI. If you want to do that, you would put the link as one of the `urls` in `pyproject.toml` (see [an example](https://packaging.python.org/en/latest/specifications/declaring-project-metadata/#example)). That can be done separately from this PR.

Either way, I think the changelog can be set up in a separate PR. But please let me know if it seems like I misunderstood.

> * Generate PyPI documentation from Github README.

That will be no problem. Actually, it's already set up that way: the documentation shown on PyPI comes from the file named in the `readme` property in `pyproject.toml`, and in this PR, the file is `README.md`. (Though now that I think about it, I should make sure the content type is set correctly. I'll take care of that.)

> commit https://github.com/bugen/pypipe/commit/5ca17c151d219c30f5b0c7ca91cdad1ac6e33f8b
>
> To run it with `pipx run`, it's necessary to create an executable file with the same name as the package, which is `pypipe-ppp`. However, it's possible to create the same executable file with a different name, like `pypipe`. In that case, it would be a good idea to add `ppp = "pypipe:main"`.
> The 'ppp' package on PyPI is not a CLI tool, but the other 'pypipe' is a CLI tool and serves a similar purpose to our 'pypipe'. So it's essential to include a note in the README to avoid confusing 'pypipe' and 'pypipe-ppp'.

Hmm, I'm not quite sure I understand. Do you mean that you do _not_ want your package to install a script named `pypipe`, so that people can install both your package and the other `pypipe` at the same time without having any conflict in script names? If so, that's no problem (though I think it will be quite unlikely to come up; your project is much more popular than that other one).

I just want to be clear: what is the set of executable files you want to create?

- `pypipe-ppp`, `pypipe`, and `ppp`
- `pypipe-ppp` and `ppp`
- `pypipe-ppp` only

By the way, this is not a big deal, but just in case you didn't know: you can comment directly on specific lines in the diff by clicking on them. (I figured I'd mention it since you said you're new to Github and open source software development.)

bugen commented on pyproject.toml:
```diff
@@ -33,6 +33,9 @@ Source = "https://github.com/bugen/pypipe"
 pypipe = "pypipe:main"
+# To support `pipx run`, there should be a console script whose name matches
+# the PyPI name of the package
+pypipe-ppp = "pypipe:main"
```
@diazona I'll try commenting on the code as you instructed :thumbsup:
Please add the following:
```
ppp = "pypipe:main"
```
Keep 'pypipe', the set of executable files is  `pypipe-ppp`, `pypipe`, and `ppp`.

diazona replied:
OK, no problem, I'll do that. I guess I'm a little confused at the purpose of having all three names for the same executable, but that doesn't stop me from implementing it. Thanks for clarifying!

bugen replied:
This is a forceful approach. What I truly want is to create only 'pypipe-ppp' and make the other two symbolic links, but from my limited research, I believe it is not possible to create symbolic links additionally using pipx. Users could create the links manually, but I don't want to add additional steps after pipx install.

diazona replied:
Ah well what I was confused about was why you want three different names at all. I mean, most packages would have only one, or at the most two, if they need an alternate name to support `pipx run`.

Anyway, you're right that it's not possible to tell pipx - or really, pip, which actually handles the installation under the hood - to make symlinks in the way you want. But I don't think it matters: the wrapper scripts that pip generates are extremely small, typically under 300 bytes. I tested it on my system and here is the script it wrote:
```python
#!/home/diazona/.local/pipx/venvs/pypipe-ppp/bin/python
# -*- coding: utf-8 -*-
import re
import sys
from pypipe import main
if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])
    sys.exit(main())
```
With such a small file, there is very little benefit to symlinking it rather than copying it.

diazona replied:
Anyway, I'll resolve this discussion since I've added `ppp` as a console script name to the `pyproject.toml` file in this PR.

bugen commented:
I mistakenly believed that I needed to manage Changelog and Release notes on PyPI and register them correctly. If Changelog and Release notes are not required on PyPI, I believe the release process could be simpler.

I manually perform the following tasks for each release:

1. Create a Git tag and push it to Github.
2. Create a Github release from the Git tag.

I want to automate the process, triggered by either 1 or 2 (whichever is easier to set up), to automatically release to PyPI and TestPyPI. This aligns with the first method you suggested in the previous comment  https://github.com/bugen/pypipe/pull/8#issuecomment-1782353481.

> When you create a Github release, the Github actions workflow will get triggered and will automatically build the packages (sdist and wheel) for later uploading. However, Github will prompt you to approve the deployment before uploading those packages to PyPI. If you don't want that release to go on PyPI, you can decline to approve it.

No approval is needed. Please set it up to run automatically from build to deployment.

>> * Generate PyPI documentation from Github README.

> That will be no problem. Actually, it's already set up that way: the documentation shown on PyPI comes from the file named in the readme property in pyproject.toml, and in this PR, the file is README.md.

 :thumbsup: That's great to hear.

Can you now complete this PR? I'm happy to provide any additional information if needed.

diazona commented:
Got it, thanks! I believe I have enough information to finish the PR now. I'll make the appropriate changes when I get time (hopefully after I get back from work today) and I'll let you know when I've done so.

bugen commented:
@diazona Thank you! There's no rush, so please proceed at your own pace without pushing yourself.

diazona commented:
Thanks for your patience! It turns out it was quite useful to have a few extra days to work on this.

I believe the PR is ready now. I added the three executable names as we discussed in the other comment thread. And I actually didn't have to change much of anything in the release workflow compared to previous versions, because what you described is how it already worked: when you create a release on Github, that will trigger the release workflow to run, and (since you don't have any deployment protection rules) it will run to completion and upload the packages to Test PyPI and PyPI without any further action from you.

I did add a test step to the release workflow, since I saw that you have automated tests now. So the release workflow will run the tests and make sure they pass before it tries to upload anything. I did it in a fairly crude way to keep things simple for now: I just copied the test steps you have in `ci.yml` and modified them a bit to run tests on the same package that will be uploaded to PyPI. This could be improved in the future, such as by using test automation tools like tox and pre-commit, and/or by using reusable workflows. But I think there's no need to get into that for this pull request.

I hope the release workflow actually works, but obviously there's no way to be sure without actually making a release. I would suggest that, after merging this PR, your next Github release should be a prerelease, something with an alpha or beta marker like `v0.3.0a0`. That way, if the workflow fails, you won't have "used up" a real version number. And you can repeatedly test it if necessary by incrementing the alpha version number, i.e. `v0.3.0a1`, `v0.3.0a2`, etc. If the release workflow does fail, then let me know and I'll be happy to help troubleshoot it.

bugen commented:
@diazona Thanks! I greatly appreciate your hard work for pypipe.

> I would suggest that, after merging this PR, your next Github release should be a prerelease, something with an alpha or beta marker like v0.3.0a0.

> But if you'd prefer to avoid making those changes to README.md until after pypipe can be installed with pip, then no problem, I'd be happy to split that commit apart into a separate PR.

As you suggest, the next release will be an alpha version, like v0.3.0a0. I'd like to confirm the successful release to PyPI before updating the README. So, please separate the commit for README from this PR or comment out the changes for now.

I recently committed the features planned for v0.3.0, so we can release v0.3.0 at any time. The time has finally come!

bugen commented:
I merged. I will try to release the alpha version in the near future.
